packages/animations/browser/src/render/transition_animation_engine.ts

```diff
@@ -290,29 +346,30 @@ export class AnimationTransitionNamespace {
     return player;
   }
 
-  deregister(name: string) {
-    this._triggers.delete(name);
-
-    this._engine.statesByElement.forEach(stateMap => stateMap.delete(name));
-
-    this._elementListeners.forEach((listeners, element) => {
-      this._elementListeners.set(element, listeners.filter(entry => {
-        return entry.name != name;
-      }));
-    });
-  }
-
-  clearElementCache(element: any) {
+  clearElementCache(element: any, context: LeaveContext|null) {
+    if (context?.recoveryState) {
+      const triggers = this._engine.statesByElement.get(element);
+      const listeners = this.elementListeners.get(element);
+      if (triggers || listeners) {
+        const details = element[REMOVAL_FLAG] as ElementAnimationState | undefined;
+        context.recoveryState.set(element, {
+          namespaceId: this.id,
```

Comment:
```suggestion
          namespaceId: details.namespaceId,
```

I've been testing this PR and this change above is required for it to work, otherwise a crash like the following will occur:

![image](https://user-images.githubusercontent.com/697707/163542414-a1a07b82-3fb4-4559-ac10-4f29b16acb03.png)

This is because the `namespaceId` is mismatched and the `restoreDescendantsStates` method will crash trying to find the trigger on the wrong namespace.

But since `namespaceId` is not nullable in `ElementRecoveryState`, `details` needs to be null checked first.

PS: I left my troubleshooting notes on Discord @JoostK

Reply:
Oh interesting, thanks for testing and investigating this issue. I would have to dive into this a bit more to fully understand the interaction of different animation namespaces here. On first glance it seems like this exposes a larger issue with how element states are cleared.

Reply:
I've been setting breakpoints and watching this, here are some related findings:

<img width="630" alt="Screenshot 2022-04-15 at 12 01 54" src="https://user-images.githubusercontent.com/697707/163550638-964e5072-3b30-443a-a7fd-bc6b03ad7d3d.png">

Notice the two namespaces, and associated elements `tournament-registration` and `tournament-calendar`.

It was `tournament-calendar` that was exhibiting the bug.

`tournament-registration` has `tournament-calendar` nested within. It appears `this.id` is the namespace of the parent in this case.

Reply:
The error/crash I attached the screenshot of was caused because it was looking for `animateIn` on the namespace that only has `animateInOut`.

With this minor change everything seems to line up properly.

Reply:
One other relevant piece of info is that this method gets called lots of times for a lot of elements, which I assume have their own animation triggers (they're from Angular Material). 

`this.id` will stay constant among all calls but `details.namespaceId` iterates among all of them.

Reply:
> this.id will stay constant (the parent) among all calls but details.namespaceId iterates among all of them (all nested children).

Indeed, and this is why I suspect there is currently an inconsistency because this logic clears the element from internal state, but the namespace that is processing the removal might not be the one associated with the element that is being removed, so I suspect we currently fail to clear data in all namespaces (perhaps the descendant namespace gets removed by itself, making this inconsistency moot).